lib/plausible/funnels.ex

```diff
@@ -0,0 +1,212 @@
+        select: {f.step, count(1)},
+        group_by: f.step
+      )
+
+    ClickhouseRepo.all(query)
+  end
+
+  defp select_funnel(db_query, funnel_definition) do
+    window_funnel_steps =
+      Enum.reduce(funnel_definition.steps, nil, fn step, acc ->
+        step_condition =
+          case step.goal do
+            %Plausible.Goal{event_name: event} when is_binary(event) ->
+              dynamic([], fragment("name = ?", ^event))
+
+            %Plausible.Goal{page_path: pathname} when is_binary(pathname) ->
+              if String.contains?(pathname, "*") do
+                regex = Plausible.Stats.Base.page_regex(pathname)
+                dynamic([], fragment("match(pathname, ?)", ^regex))
+              else
+                dynamic([], fragment("pathname = ?", ^pathname))
+              end
+          end
+
+        if acc do
+          dynamic([q], fragment("?, ?", ^acc, ^step_condition))
+        else
+          dynamic([q], fragment("?", ^step_condition))
+        end
+      end)
+
+    dynamic_window_funnel =
+      dynamic(
+        [q],
+        fragment("windowFunnel(?)(timestamp, ?)", @funnel_window_duration, ^window_funnel_steps)
+      )
+
+    from(
+      q in db_query,
+      select_merge:
+        ^%{
+          step: dynamic_window_funnel
+        }
+    )
+  end
+
+  defp backfill_steps(funnel_result, funnel) do
+    # Directly from ClickHouse we only get {step_idx(), visitor_count()} tuples.
+    # but no totals including previous steps are aggregated.
+    # Hence we need to perform the appropriate backfill
+    # and also calculate dropoff and conversion rate for each step.
+    funnel_result = Enum.into(funnel_result, %{})
+    max_step = Enum.max_by(funnel.steps, & &1.step_order).step_order
+
+    funnel
+    |> Map.fetch!(:steps)
+    |> Enum.reduce({nil, nil, []}, fn step, {total_visitors, visitors_at_previous, acc} ->
+      # first step contains the total number of all visitors qualifying for the funnel,
+      # with each subsequent step needing to accumulate sum of the previous one(s)
+      visitors_at_step =
+        step.step_order..max_step
+        |> Enum.map(&Map.get(funnel_result, &1, 0))
+        |> Enum.sum()
+
+      # accumulate current_visitors for the next iteration
+      current_visitors = visitors_at_step
+
+      # First step contains the total number of visitors that we base percentage dropoff on
+      total_visitors =
+        total_visitors ||
+          current_visitors
+
+      # Dropoff is 0 for the first step, otherwise we subtract current from previous
+      dropoff = if visitors_at_previous, do: visitors_at_previous - current_visitors, else: 0
+
+      conversion_rate =
+        if current_visitors == 0 or total_visitors == 0 do
+          "0.00"
+        else
+          # XXX: Talk to Vini
+          (current_visitors / total_visitors * 100)
+          |> Decimal.from_float()
+          |> Decimal.round(2)
+          |> Decimal.to_string()
```

Comment:
Just a nitpick to avoid converting from floats. Feel free to ignore.

```suggestion
          Decimal.div(current_visitors, total_visitors)
          |> Decimal.mult(100)
          |> Decimal.round(2)
          |> Decimal.to_string()
```

Reply:
Ah cool